### PR TITLE
nixos/xmonad: make a wrapper

### PR DESCRIPTION
**Pull request by abbradar**
This allows one to use "xmonad --recompile" without problems.

### CODE REVIEW

**Comment by abbradar** [comment]
Sorry for taking a long time; removed version and a patch in question.

**Comment by peti** [comment]
This looks good to me. :+1:

**Comment by eduarrrd** [comment]
What are the outstanding issues with this PR as of now?

**Comment by MP2E** [comment]
I've reviewed this, and it works well with haskellPackages. However, it seems to break xmonad on haskellngPackages, mainly due to there being no wrapper derivation in haskellngPackages.

I think with haskell-ng support, this PR can finally be merged in to master.

**Comment by abbradar** [comment]
What is our deprecation policy with regards to `xmonad`, `taffybar` and such in `haskellPackages`? I have removed old versions from the set altogether, but on the second thought we may want to leave them in case someone uses them with an old set.

**Comment by peti** [comment]
Removing those packages from `haskellPackages` is probably a bit extreme. I would keep them around and assume that everyone who cares about Haskell in Nixpkgs has heard of the ng package set by now and knows what he's doing when using the old versions. We'll throw all the old stuff out when GHC 7.10.x becomes our default compiler.

**Comment by abbradar** [comment]
Let it be so, then! I have the same opinion after more thought now; question popped in my head when I've thought that `xmonad` and friends are probably used only in NixOS module, but that is not necessarily true.

**Comment by abbradar** [comment]
New version, using `haskell-ng`. I've moved needed patch to `haskell-modules` as it's done for `ghc-paths`; if that's undesirable, I can leave it where it originally was.

File: pkgs/development/haskell-modules/hackage-packages.nix [hunk 1]
```diff
@@ -122375,6 +122375,7 @@ self: {
          base containers directory extensible-exceptions filepath mtl
          process unix utf8-string X11
        ];
+       patches = [ ./xmonad-nix.patch ];
```
Comment (peti) [inline pkgs/development/haskell-modules/hackage-packages.nix]: Please don't edit `hackage-packages.nix`. This patch must be added by an override in `configuration-common.nix` to persist through an update.

Comment (abbradar) [inline pkgs/development/haskell-modules/hackage-packages.nix]: No problem! Perhaps `ghc-paths` `patches` entry is a bug then, though.

**Comment by abbradar** [comment]
Fixed. BTW, @peti, thanks for this great new infrastructure!

File: pkgs/development/haskell-modules/hackage-packages.nix (hunk 1 above)
Comment (peti) [inline pkgs/development/haskell-modules/hackage-packages.nix]: The `patch` attribute comes from this code: https://github.com/NixOS/cabal2nix/blob/master/src/Cabal2Nix/PostProcess.hs#L44. I agree that this is a bit of a bug; I haven't come around to cleaning that up yet.

**Comment by abbradar** [comment]
Hm, do we want to save compatibility there? Users would just need to rename packages in their configurations. I'm asking because renaming to the old scheme would break `haskellngPackages` by default and also IIRC things like `haskell-ng.packages.ghc784` wouldn't work (they don't have old attribute names at all).

**Comment by MP2E** [comment]
Fair enough, I didn't realize it broke things on the Haskell-NG end.

Thanks! Now we finally have an xmonad that can rebuild :+1:

**Pull request merged**